datacube_stats/main.py

```diff
@@ -940,5 +965,168 @@ def __call__(self, product, time, group_by):
+    """
+    Create a :class:`Tile` which can be used by :class:`Generate_Polygon_Tasks` to later load the
+    filtered datasets.
+
+    """
+    def __init__(self, index, filter_product, storage):
+        self.dc = Datacube(index=index)
+        self.filter_product = filter_product
+        self.storage = storage
+
+    def __call__(self, product, time, group_by, filter_time, geopoly, poly_dates):
+        # Find the sources for each layer
+        if poly_dates == 1:
+            return self.dc
+        datasets = self.dc.find_datasets(product=product, time=time, geopolygon=geopoly, group_by=group_by)
+        fil_datasets = list()
+        # check for appropriate products to populate required date list
+        if self.filter_product.get('method') == 'by_hydrological_months':
+            dt = filter_time
+        else:  # Fot ITEM/low/high products
+            dt = [dd[0] for dd in filter_time]
+        # get the date list out of dt string and compare within the epoch
+        for ds in datasets:
+            if ds.time.begin.date().strftime("%Y-%m-%d") in dt:
+                fil_datasets.append(ds)
+        if len(fil_datasets) == 0:
+            _LOG.info("No matched for product %s on %s", product, str(filter_time))
+            return None
+        datasets = fil_datasets
+        output_crs = CRS(self.storage['crs'])
+        output_resolution = [self.storage['resolution'][dim] for dim in output_crs.dimensions]
+        group_by = query_group_by(group_by=group_by)
+        sources = self.dc.group_datasets(datasets, group_by)
+        geobox = GeoBox.from_geopolygon(geopoly, resolution=output_resolution)
+        return Tile(sources, geobox)
+
+
+# def Generate_Polygon_Tasks(index, sources_spec, date_ranges, filter_product, input_region, storage):
+class Generate_Polygon_Tasks(object):
+    """
+    Make stats tasks for a defined spatial region, that doesn't fit into a standard grid.
+    :param index: database index
+    :param input_region: dictionary of query parameters defining the target input region. Usually
+                         x/y spatial boundaries.
+    :return:
+    """
+    def __init__(self, input_region, filter_product, storage):
+        self.input_region = input_region
+        self.filter_product = filter_product
+        self.storage = storage
+        self.tile_index = ()
+        self.filter_time = ""
+        self.feature = {}
+
+    def __call__(self, index, sources_spec, date_ranges):
+        """
+        Generate the required tasks through time and across a feature ID of a polygon.
+
+        feature id is passed through filter_product and is processed after filtering the relevant dates
+        that will both result in only datasets covering the poly.
+
+        :param index: Datacube Index
+        :return:
+        """
+        if self.filter_product['args'].get('feature_id') is None:
+            _LOG.info("feature_id must be provided.  EXITING ...")
+            return
```

Comment:
I think `return` here will cause a crash, since __call__ should `yield`ing instead. It also might end up masking the error, not entirely sure.

Reply:
And if it is actually existing, maybe raise an exception or otherwise better deal with the error case.